sktime/transformations/base.py

```diff
@@ -48,93 +104,567 @@ class BaseTransformer(BaseEstimator):
+
+        valid, msg, metadata = check_is(
+            X, mtype=self.ALLOWED_INPUT_MTYPES, return_metadata=True, var_name="X"
+        )
+        if not valid:
+            raise ValueError(msg)
+
+        # checking X
+        enforce_univariate = self.get_tag("univariate-only")
+        if enforce_univariate and not metadata["is_univariate"]:
+            raise ValueError("X must be univariate but is not")
+
+        # retrieve mtypes/scitypes of all objects
+        #########################################
+
+        X_input_mtype = mtype(X)
+        X_input_scitype = mtype_to_scitype(X_input_mtype)
+        y_input_mtype = mtype(y)
+        y_input_scitype = mtype_to_scitype(y_input_mtype)
+
+        X_inner_mtype = self.get_tag("X_inner_mtype")
+        if not isinstance(X_inner_mtype, list):
+            X_inner_mtype = [X_inner_mtype]
+        X_inner_scitypes = list(set([mtype_to_scitype(mt) for mt in X_inner_mtype]))
+
+        y_inner_mtype = self.get_tag("y_inner_mtype")
+        if not isinstance(y_inner_mtype, list):
+            y_inner_mtype = [y_inner_mtype]
+        # y_inner_scitypes = list(set([mtype_to_scitype(mt) for mt in y_inner_mtype]))
+
+        # treating Series vs Panel conversion for X
+        ###########################################
+
+        # there are three cases to treat:
+        # 1. if the internal _fit supports X's scitype, move on to mtype conversion
+        # 2. internal only has Panel but X is Series: consider X as one-instance Panel
+        # 3. internal only has Series but X is Panel:  loop over instances
+        #     currently this is enabled by conversion to df-list mtype
+        #     and this does not support y (unclear what should happen here)
+
+        # 1. nothing to do - simply don't enter any of the ifs below
+
+        # 2. internal only has Panel but X is Series: consider X as one-instance Panel
+        if X_input_scitype == "Series" and "Series" not in X_inner_scitypes:
+            X = convert_Series_to_Panel(X)
+
+        # 3. internal only has Series but X is Panel: loop over instances
+        elif X_input_scitype == "Panel" and "Panel" not in X_inner_scitypes:
+            if y is not None:
+                raise ValueError(
+                    "no default behaviour if _fit does not support Panel, "
+                    " but X is Panel and y is not None"
+                )
+            X = convert_to(X, to_type="df-list", as_scitype="Panel")
+            # this fits one transformer per instance
+            self.transformers_ = [clone(self).fit(Xi) for Xi in X]
+            # recurse and leave function - recursion does input checks/conversion
+            # also set is_fitted flag to True since we leave function here
+            self._is_fitted = True
+            return self
+
+        X_mtype = mtype(X)
+        X_scitype = mtype_to_scitype(X_mtype)
+
+        assert X_scitype in X_inner_scitypes, "conversion of X to X_inner unsuccessful"
```

Comment:
Do not use assets like this. Depending on compiler flags assert can be disabled (i.e. asserts only work in debug builds and wont work in release builds).

Reply:
replaced by a raise